Update Caching and Security with Cloud SDK for Version 3.0

MatKuhr opened the pull request:
Updates the following Tutorials for Version 3.0 of the Cloud SDK:

- [Introduce caching to your application](https://developers-qa.sap.com/tutorials/s4sdk-caching.html)
- [Secure your Application on Cloud Foundry](https://developers-qa.sap.com/tutorials/s4sdk-secure-cloudfoundry.html)

sacnl commented on tutorials/s4sdk-caching/s4sdk-caching.md:
```diff
@@ -51,268 +56,136 @@ If each of these options apply to your use case, then we highly recommend the ca
 [ACCORDION-BEGIN [Step 3: ](Caching Command with SAP Cloud SDK)]
 
-The command cache allows parallel execution and asynchronous computation for efficient programming practices. Stored information is organized as a local key-value store. For each unique key the same response is expected in case of a "cache hit". Otherwise the cache response will be computed.
+The caching command allows parallel execution and asynchronous computation for efficient programming practices.
```
The term caching command (also in the heading) no longer applies in version 3. Also, parallelization is not in focus here.

sacnl commented on tutorials/s4sdk-caching/s4sdk-caching.md:
```diff
@@ -51,268 +56,136 @@ If each of these options apply to your use case, then we highly recommend the ca
+2. Declare the parameters that need to be stored together with the cached data
 
-import com.sap.cloud.sdk.s4hana.datamodel.odata.namespaces.businesspartner.BusinessPartner;
+The first step is obviously necessary, since we want to set up a cache that is meant to store data only a short amount of time to optimize execution time. How long you want to keep data cached depends on the data that you receive. How fast do you expect the information to be outdated? How frequently will the data be accessed?
```
Why does storing a short amount of time optimize execution time? I would rather speak of a trade-off between data being up-to-date and fast response, which depends on the use case at hand.

sacnl commented on tutorials/s4sdk-caching/s4sdk-caching.md:
```diff
@@ -51,268 +56,136 @@ If each of these options apply to your use case, then we highly recommend the ca
+
+Now that we have a working command with caching functionality we also have to adapt our test. Recall the test we prepared to check our resilient command falls back to an empty list in case of failure. Note that this behavior has now changed slightly.
+
+If our servlet got the desired result cached from a previous call, and the ERP system is temporarily not available, our cache will still return the data. But the test expects the result to be empty in that case. In order to account for this behavior and to see if our cache is working as expected let's adapt the test to account for caching. Replace the `testWithFallback` test with the following co ...
```
Why replace the test instead of adding a new one and adapt the existing one (in the next section)? The fallback functionality still needs to be tested.

sacnl commented on tutorials/s4sdk-caching/s4sdk-caching.md:
```diff
@@ -51,268 +56,136 @@ If each of these options apply to your use case, then we highly recommend the ca
-- As before, in case the `runCachable` procedure fails in any way, we provide a fallback solution. In this example we return an empty list object
 
-To use the new class in the servlet, just read a parameter from the "`request`" input variable and use it for instantiating this `parameterized` command class. Or you could create different servlet endpoints for persons (category "1"), organizations (2) and groups (3), which supply the corresponding category to the newly created `parameterized` command. Feel free to try it out!
+Executed on it's own, this test will still pass as it did before. However, if `testService` is run prior to this test the result will be cached and `testWithFallback` will fail since the cached data is returned instead of falling back to an empty list. By changing the test we avoided such issues since our new test does not interfere with other tests. While being sufficient for this tutorial, in a ...
```
One could also explicitly invalidate the cache prior to executing the fallback test (with drawbacks for parallel execution of tests)

MatKuhr replied:
discussed offline

sacnl commented on tutorials/s4sdk-secure-cloudfoundry/s4sdk-secure-cloudfoundry.md:
````diff
@@ -90,168 +69,148 @@ This alternative approach requires that you have npm installed on your machine.
+
+    ```yaml
+    routes:
+      - route: firstapp-<SUBACCOUNT>.cfapps.<REGION_IDENTIFIER>.hana.ondemand.com
````
Thinking about it again, I don't think that it is necessary to adapt the route of the backend. The tenant identification is only necessary for the approuter route. As long as the correct URL is entered in destinations below, all should be fine.

MatKuhr replied:
I believe you are correct. However, I am not sure how random-route behaves when redeploying, especially when you do that by deleting the app in the cockpit and pushing it again. The point here is more to have a stable, unique URL. I'll test if it works with a random route even after redeploying.

sacnl commented on tutorials/s4sdk-secure-cloudfoundry/s4sdk-secure-cloudfoundry.md:
```diff
@@ -90,168 +69,148 @@ This alternative approach requires that you have npm installed on your machine.
 
-[DONE]
-[ACCORDION-END]
+    Replace `<SUBACCOUNT>` with your actual account, which is the one you use in the CF CLI to login to Cloud Foundry. With this step we determine a route instead of generating a random route. This is important since we will configure the app router for exactly this destination. Further change the `<REGION_IDENTIFIER>` to match your region (e.g. `eu10`). If you are unsure about your region, head  ...
```
See above, move this down, but add an explanation on how to identify the value for `<SUBACCOUNT>`. This is not self-evident for readers. They need to go into the SCP cockpit and look for the subdomain associated with the subaccount, so a screenshot will be helpful.

MatKuhr replied:
AFAIK the subaccount is the one required when logging into CF with `cf login`. Given the instructions in [creating the sample app](https://developers.sap.com/tutorials/s4sdk-cloud-foundry-sample-application.html) I feel it should be clear. Or am I missing something here?

sacnl commented on tutorials/s4sdk-secure-cloudfoundry/s4sdk-secure-cloudfoundry.md:
```diff
@@ -90,168 +69,148 @@ This alternative approach requires that you have npm installed on your machine.
+      - In `destinations` replace `<APPLICATION_URL>` with the actual URL of your previously deployed app.
 
-**Destinations** is a variable that declares the internal routes from the App Router to the underlying backend microservices. As we only have one hello world microservice yet, we define only one destination called app-destination here. This app-destination is referenced by the previously created `xs-app`.json file.
+**Note**: It is mandatory that your app URL conforms the format `<APP_NAME>-<SUBACCOUNT>.cfapps.<REGION_IDENTIFIER>.hana.ondemand.com`. The authentication mechanism will identify you by pulling the user-id (`SUBACCOUNT`) out of the subdomain.
```
See above. It is not the user-id that gets pulled out, but the subdomain which is then mapped to the tenant-id.

MatKuhr replied:
I believe the wording is correct. The subdomain is `approuter-UID`, the pattern `approuter-(.*).cfapps...` handles extracting the UID = tenant id from the subdomain.

sacnl commented on tutorials/s4sdk-secure-cloudfoundry/s4sdk-secure-cloudfoundry.md:
```diff
@@ -90,168 +69,148 @@ This alternative approach requires that you have npm installed on your machine.
+### Understanding the AppRouter's `manifest.yml` and `xs-app.json`
 
-![XSUAA service](Figure3-1-1.png)
+The `TENANT_HOST_PATTERN` is a variable that declares the pattern how multiple tenants in the URL are identified and handled. During runtime, the App Router will match the incoming host against the pattern and tries to extract the regular expression from the pattern. It will use the matched string to delegate authentication to the respective XSUAA tenant (in SAP Cloud Platform the tenant ID corre ...
```
While tenant ID = subaccount ID holds, this is not the subdomain used here, so tenant ID is not directly related to the URL pattern.

sacnl commented on tutorials/s4sdk-secure-cloudfoundry/s4sdk-secure-cloudfoundry.md:
````diff
@@ -90,168 +69,148 @@ This alternative approach requires that you have npm installed on your machine.
 ```
 
-`7.` Before we deploy your folder structure should look similar to this:
+Then deploy the app router using the following (with the appropriate API endpoint of your Cloud Foundry region):
````
```suggestion
After you have created the XSUAA service instance, deploy the app router using the following (with the appropriate API endpoint of your Cloud Foundry region):
```

sacnl commented on tutorials/s4sdk-secure-cloudfoundry/s4sdk-secure-cloudfoundry.md:
```diff
@@ -479,68 +328,69 @@ If you have previously exposed the backing service directly to the end user, you
-[ACCORDION-BEGIN [Step 16: ](Appendix)]
+[ACCORDION-BEGIN [Step 9: ](Appendix)]
 ### Understanding Roles, Role Collections and Scopes
 The following picture explains how the various concepts are related to each other.
```
There is a picture missing here.

sacnl commented on tutorials/s4sdk-secure-cloudfoundry/s4sdk-secure-cloudfoundry.md:
```diff
@@ -565,33 +415,31 @@ Afterwards you may use `https://jwt.io/` to decode the token. **Note:** You shou
 To make this happen, the IdP and the service provider (SP) have to exchange security metadata, i.e., the IdP has to import the metadata of the SP and vice versa.
 
-You can retrieve the metadata from your XSUAA tenant by following the pattern **https://<tenantID>.`authentication.eu10.hana.ondemand.com`/`saml`/metadata**. In my case, that would be `https://p1942765239trial.authentication.eu10.hana.ondemand.com/saml/metadata`. This downloads the `metatdata` to your local computer.
+You can retrieve the metadata from your XSUAA tenant by following the pattern `https://<tenantID>.authentication.eu10.hana.ondemand.com/saml/metadata`. In my case, that would be `https://p1942765239trial.authentication.eu10.hana.ondemand.com/saml/metadata`. This downloads the metadata to your local computer.
```
Replace `<tenantID>` with subdomain as above for consistency.

sacnl commented on tutorials/s4sdk-secure-cloudfoundry/s4sdk-secure-cloudfoundry.md:
```diff
@@ -90,168 +69,148 @@ This alternative approach requires that you have npm installed on your machine.
-[ACCORDION-END]
-
-[ACCORDION-BEGIN [Step 6: ](Install XS Security libs to your local Maven repository:)]
-The first step is to get some additional Java libs from Service Marketplace. To get them, do the following steps:
```
This is still required, isn't it, if you use the dependencies listed in the pom below. For example, `com.sap.xs2.security` is not on Maven Central.

However, the setup should be easier with the approach described here: [with SAP buildpack](https://github.com/SAP/cloud-security-xsuaa-integration/tree/master/samples/sap-java-buildpack-api-usage) or here: [with Spring security](https://github.com/SAP/cloud-security-xsuaa-integration/tree/master/spring-xsuaa)

MatKuhr commented:
@sacnl I changed the wording and explanation around the sub-account, subdomain and tenant topic. Now it should be more clear what is happening. Also, I added the installation step for the java libs again and removed the optional step of renaming the URL in order to reduce complexity.

pacaceres commented on tutorials/s4sdk-secure-cloudfoundry/s4sdk-secure-cloudfoundry.md:
```diff
@@ -90,168 +69,150 @@ This alternative approach requires that you have npm installed on your machine.
+      1. Replace `<subdomain>` with your subdomain. You will find your subdomain in the CF cockpit by heading to the overview page of your sub-account:
 
-2. Within <destLocation> create a new **`manifest.yml`** file for the AppRouter microservice with the following content.
+          ![Subomain and Tenant ID in the CF Cockpit](Figure-3-1.png)
```
This image needs to show the Subdomain, but it appears to be showing part of a tenant ID.

MatKuhr replied:
The subdomain and most parts of the tenant ID are blurred out in order to not give away account details.
